File: packages/android-client/android/src/main/java/voltra/widget/server/WidgetServerEtagStore.kt
package voltra.widget.server

import android.content.Context

/**
 * The ETag from the last `200`, stored with the URL it came from.
 *
 * Keeping the URL alongside it is what makes `If-None-Match` safe once the app can change the URL
 * at runtime: an ETag minted by one endpoint says nothing about another, and sending it could
 * produce a `304` that leaves the widget showing the previous endpoint's data forever.
 */
class WidgetServerEtagStore(
    context: Context,
) {
    private val preferences = context.applicationContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    /** The stored ETag, but only if it was minted by [url]. */
    fun etag(
        scope: WidgetScope,
        url: String?,
    ): String? {
        if (url == null) return null
        if (preferences.getString(urlKey(scope), null) != url) return null

        return preferences.getString(etagKey(scope), null)
    }

    fun put(
        scope: WidgetScope,
        url: String,
        etag: String?,
    ) {
        preferences
            .edit()
            .apply {
                if (etag == null) {
                    remove(etagKey(scope))
                    remove(urlKey(scope))
                } else {
                    putString(etagKey(scope), etag)
                    putString(urlKey(scope), url)
                }
            }.apply()
    }

    fun clear(scope: WidgetScope) {
        preferences
            .edit()
            .remove(etagKey(scope))
            .remove(urlKey(scope))
            .apply()
    }

    private fun etagKey(scope: WidgetScope) = "etag.${scope.storageKey}"

    private fun urlKey(scope: WidgetScope) = "etag_url.${scope.storageKey}"

    companion object {
        private const val PREFERENCES_NAME = "voltra_widget_server_etags"
    }
}

File: packages/ios-client/ios/shared/WidgetServer/WidgetServerEtagStore.swift
import Foundation

/// The ETag from the last `200`, stored with the URL it came from.
///
/// Keeping the URL alongside it is what makes `If-None-Match` safe once the app can change the URL
/// at runtime: an ETag minted by one endpoint says nothing about another, and sending it could
/// produce a `304` that leaves the widget showing the previous endpoint's data forever.
///
/// Lives in the App Group so the widget extension and the app agree on it. Without an App Group
/// there is nowhere shared to put it, and the widget simply revalidates nothing.
public enum WidgetServerEtagStore {
  /// The stored ETag, but only if it was minted by `url`.
  public static func etag(for scope: WidgetScope, url: String?) -> String? {
    guard let url, let defaults = defaults(),
          defaults.string(forKey: urlKey(scope)) == url
    else {
      return nil
    }

    return defaults.string(forKey: etagKey(scope))
  }

  public static func put(_ etag: String?, for scope: WidgetScope, url: String) {
    guard let defaults = defaults() else { return }

    guard let etag else {
      clear(scope)
      return
    }

    defaults.set(etag, forKey: etagKey(scope))
    defaults.set(url, forKey: urlKey(scope))
  }

  public static func clear(_ scope: WidgetScope) {
    guard let defaults = defaults() else { return }

    defaults.removeObject(forKey: etagKey(scope))
    defaults.removeObject(forKey: urlKey(scope))
  }

  private static func defaults() -> UserDefaults? {
    guard let group = VoltraConfig.groupIdentifier() else { return nil }

    return UserDefaults(suiteName: group)
  }

  private static func etagKey(_ scope: WidgetScope) -> String {
    "Voltra_WidgetServerEtag_\(scope.storageKey)"
  }

  private static func urlKey(_ scope: WidgetScope) -> String {
    "Voltra_WidgetServerEtagUrl_\(scope.storageKey)"
  }
}
